docs/adr/0020-application-architecture.md

```diff
@@ -0,0 +1,110 @@
+- Time consuming to setup boiler-plate for components, as we have no tooling to work with the current ad-hoc application architectures.
+- Our ad-hoc architecture frequently mixes concerns such as having business logic mixed in with request handling logic, and has other warts from its evolution over time vs. being planned up front.
+- New back-end services evolve differently resulting in more ad-hoc application architectures to learn.
+- Shared components in `fxa-shared` can't rely on basic object lifecycles or setup approaches as they may be used in multiple different ad-hoc application architectures.
+
+Not choosing an application framework means that we have choosen to make ad-hoc application architectures which will continue to exhibit the problems above.
+
+It is assumed that the four newest FxA back-end services (admin-server, support-panel, event-broker, gql-api) will be switched to the chosen approach for consistency.
+
+## Decision Drivers
+
+- Documented application architecture.
+- Tooling that reduces boiler-plate and creates consistent code architecture.
+- Modern paradigms to ease creation of global objects and their testability, such as DI.
+- Training materials so that new developers can understand and work on application features easily.
+- Handles situations we need such as:
+  - Exception handling
+  - Validation
+  - Service instantiation (DI, etc)
+  - Authentication
+  - RESTful API's (Swagger/OpenAPI a bonus)
+  - GraphQL
+  - Use of components/services easily in a script
+
+## Considered Options
+
+- Evolve/improve our ad-hoc application architecture with docs, tools, etc.
+- Use an existing well-documented and popular application framework such as:
+  - NestJS
+  - Adonis
+  - LoopBack
+
+## Decision Outcome
+
+### Positive Consequences
+
+### Negative Consequences
+
+## Pros and Cons of the Options
+
+### Evolve/improve our ad-hoc application architecture with docs, tools, etc.
+
+Our existing ad-hoc structure typically involves a few chosen libraries and a basic directory structure. Error handling differs in several of the back-end services, while validation is always provided by the `joi` library. To document our ad-hoc application architecture we would first need to make clear choices about what libraries we use and how their use should be structured in the application. We could then write tooling and documentation around these choices.
+
+- Pros:
+  - Could be easier to convert our existing backend services as they will likely be closer in libs/structure to the choosen structure if desired.
+  - Our choice of exactly what our application architectures look like and what libraries are used.
+- Cons:
+  - Creating even a portion of the documentation, tooling, and training material as already exists for any of the other three options is an incredible amount of effort/work.
+  - We will likely end up copying many of the features/layout/tools that already exist with communities for the other options.
```

Comment:
I broadly endorse the list of cons here, for whatever that's worth; I think the process of trying to go through and clean it up to turn into a documented application architecture would look pretty similar to slowly evolving it towards some existing tool, so might as well head towards something with an existing ecosystem around it.